gcc/cp/contracts.cc

```diff
@@ -1613,61 +1618,208 @@ build_contract_function_decls (tree d)
   set_contract_functions (d, pre, post);
 }
 
-static const char *
-get_contract_level_name (tree contract)
+/* Build a layout-compatible internal version of source location __impl type.  */
+static tree
+get_pseudo_source_location_impl_type ()
 {
-  if (CONTRACT_LITERAL_MODE_P (contract))
-    return "";
-  if (tree mode = CONTRACT_MODE (contract))
-    if (tree level = TREE_VALUE (mode))
-      return IDENTIFIER_POINTER (level);
-  return "default";
+  if (!pseudo_source_location_impl_type)
```

Comment:
[nit, not  really a review] you could save a bunch of indenting by doing an early return here if pseudo_source_location_impl_type is already built (and in the following cases too) ...

Reply:
done